### DISCUSSION TITLE
Expose @prisma/client to NPM package

### DISCUSSION THREAD

**Discussion by DanielSinclair** [prisma/prisma#7195] [Q&A]
I'd like to share my Prisma schema and client across different microservices, so that they can share the underlying database. Absent a monorepo, the easiest way to do this in node would be to create a new package where the underlying prisma schema and client lives, then to expose that via an export. 

I've run into a couple of different blockers that have prevented this strategy. To reduce potential issues regarding linking and CI environments, I've used custom generation with the following schema. The package is compiled and bundled via a GitHub Action runner. I've added the various binary targets in the case that the compile environment was different than runtime. It is generated to `./generated/client`

```prisma
generator client {
  provider          = "prisma-client-js"
  output             = "./generated/client"
  binaryTargets = ["native", "darwin", "debian-openssl-1.1.x", "linux-musl"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URI")
}
```

The package consists of the following. It is is compiled via babel with `babel index.js -d lib && cp -R generated lib/generated`
```ts
import { PrismaClient } from './generated/client/index'
const prisma = new PrismaClient()
export default prisma
```

The babel.config.json
```json
{
  "presets": ["@babel/preset-env"]
}
```

What the lib looks like upon generation and compilation: 
<img width="343" alt="Screen Shot 2021-05-20 at 10 05 09 PM" src="https://user-images.githubusercontent.com/4412473/119071314-92eb9b00-b9b7-11eb-93ab-14b4c3b41bcc.png">

Upon runtime in a simple ncc environment, I receive the following error:
```
Invalid `prisma.account.findMany()` invocation:

  Query engine exited with code 101

thread 'main' panicked at 'Could not open datamodel file "/private/var/folders/3l/9mhb9m_94t7_c1lg__kkwr9h0000gn/T/ea73cbe39dfea15b66fca71239b7300f/schema.prisma"', query-engine/query-engine/src/opt.rs:185:53
stack backtrace:
   0: _rust_begin_unwind
   1: std::panicking::begin_panic_fmt
   2: query_engine::opt::load_datamodel_file::{{closure}}
   3: structopt::StructOpt::from_args
   4: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
   5: std::thread::local::LocalKey<T>::with
   6: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
   7: async_io::driver::block_on
   8: async_global_executor::reactor::block_on
   9: std::thread::local::LocalKey<T>::with
  10: async_std::task::builder::Builder::blocking
  11: query_engine::main
```

**Comment by ryands17** [comment]
@DanielSinclair 👋 

I don't know if this article [might](https://isidoro-ferreiro.medium.com/share-your-prisma-client-across-projects-44d1c7aca6fd) help your use case, if not I would suggest opening an [issue](https://github.com/prisma/prisma/issues/new?assignees=&labels=kind%2Fbug&template=bug_report.yml) here with an example repo so that we can debug this :)

**Reply by dakshshah96** [reply]
Hey, is this article still the suggested way to accomplish this? I have a monorepo with 3 servers all connecting to the same database, and right now we have to repeat the Prisma steps in all 3 places each time there is a database schema change.

**Reply by alvis** [reply]
I don't think so. The `.prisma` folder is generated under `node_modules` which is not packaged via npm.

**Reply by nicofishman** [reply]
Any updates on this? I'm having trouble adding the .prisma/client/index.d.ts file to my package

**Reply by eliliam** [reply]
I'm hitting this same thing now too, any official advice on this route? Seems like it would be a more common use case, but it's hard to find anyone else asking about it .

**Reply by JavascriptMick** [reply]
Hi @eliliam and gang,  I muddled through this one tonight.  This seems to be working for me.

Key things: include the schema.prisma file in the package as well as the dist folder (see the files bit).... post install script to do the generate.... post install uses npx so I don't need prisma installed in the consuming repo....prisma is a dev dependency (not installed along with package), @prisma/client is a normal dependency and *is* installed with the package.

here is my package.json (I cut out stuff that was irrelevant)
```
{
  "name": "my_cool_business_service",
  "version": "0.0.11",
  "files": ["dist", "prisma/schema.prisma"],
  "main": "./dist/index.js",
  "exports": {
    "require": "./dist/index.js",
    "import": "./dist/index.js"
  },
  "types": "./dist/index.d.ts",
  "license": "UNLICENCED",
  "private": true,
  "prisma": {
    "seed": "ts-node prisma/seed.ts"
  },
  "scripts": {
    "build": "tsc",
    "lint": "eslint . --ext .ts",
    "postinstall": "npx prisma@5.15.0 generate",
    "test": "vitest run",
    "test:watch": "vitest"
  },
  "devDependencies": {
    "prisma": "^5.10.2"
  },
  "dependencies": {
    "@prisma/client": "^5.15.0"
  }
}
```

and just in case, here is my tsconfig
```
{
  "compilerOptions": {
    "target": "ESNext",
    "module": "ESNext",
    "moduleResolution": "node",
    "strict": true,
    "esModuleInterop": true,
    "skipLibCheck": true,
    "forceConsistentCasingInFileNames": true,
    "outDir": "./dist",
    "rootDir": "./src",
    "declaration": true,
    "resolveJsonModule": true,
    "allowSyntheticDefaultImports": true
  },
  "include": ["src/**/*.ts"],
  "exclude": ["node_modules", "dist"]
}
```

**Reply by eliliam** [reply]
@JavascriptMick interesting, I'll have to play around with something along the same lines. Thanks for posting those details, I think you're on to a potential solution for this need.
